sample_size/sample_size_calculator.py

```diff
@@ -0,0 +1,80 @@
+from typing import List
+from typing import Union
+
+import numpy as np
+
+from sample_size.metrics import BooleanMetric
+from sample_size.metrics import NumericMetric
+from sample_size.metrics import RatioMetric
+
+DEFAULT_ALPHA = 0.05
+DEFAULT_POWER = 0.8
+
+
+class SampleSizeCalculator:
+    """This class is to calculate sample size based on variable type
+
+    Attributes:
+    alpha: statistical significance
+    power: statistical power
+
+    """
+
+    def __init__(self, alpha: float = DEFAULT_ALPHA, power: float = DEFAULT_POWER):
+        self.alpha = alpha
+        self.power = power
+        self.boolean_metrics: List[BooleanMetric] = []
+        self.numeric_metrics: List[NumericMetric] = []
+        self.ratio_metrics: List[RatioMetric] = []
+
+    def register_bool_metric(self, probability: float, mde: float) -> None:
+        metric = BooleanMetric(probability, mde)
+        self.boolean_metrics.append(metric)
+
+    def register_numeric_metric(self, variance: float, mde: float) -> None:
+        metric = NumericMetric(variance, mde)
+        self.numeric_metrics.append(metric)
+
+    def register_ratio_metric(
+        self,
+        numerator_mean: float,
+        numerator_variance: float,
+        denominator_mean: float,
+        denominator_variance: float,
+        covariance: float,
+        mde: float,
+    ) -> None:
+        metric = RatioMetric(
+            numerator_mean, numerator_variance, denominator_mean, denominator_variance, covariance, mde
+        )
+        self.ratio_metrics.append(metric)
+
+    def _get_single_sample_size(self, metric: Union[BooleanMetric, NumericMetric, RatioMetric]) -> float:
+        effect_size = metric.mde / float(np.sqrt(metric.variance))
+        power_analysis = metric.default_power_analysis_type()
+        sample_size = int(
+            power_analysis.solve_power(
+                effect_size=effect_size,
+                alpha=self.alpha,
+                power=self.power,
+                ratio=1,
+                alternative="two-sided",
+            )
+        )
+        return sample_size
+
+    def get_overall_sample_size(self) -> float:
+        # Supports the sample size calculation for single metric now.
+        # The current structure is set up to support multiple metrics in the future.
+        sample_size = float("nan")
+        if self.boolean_metrics:
```

Comment:
since `default_power_analysis_type` is under each variable class now, it looks like we can actually just have single variable `self.metrics`. but not sure if the current setting (three `self.{type}_metric`s) will be easier for multiple metrics correction later on.
Same idea as the comment that @vjohnson1-godaddy has https://github.com/godaddy/sample-size/pull/4#discussion_r759654663

Reply:
Yeah it depends on whether we need to apply multiple metric correction based on the metric type. How about we keep it like this and refactor (if necessary) after the research is completed?

Reply:
sounds good, can you add a Note like ` consider having only one self.metrics` , thanks

Reply:
By "multiple metric correction" do we mean false discovery rate adjustment?

If so, I am _fairly_ confident the _type_ of metric won't be material to the adjustment (however open to designing around this just in case).

Perhaps we could represent the metrics as a dictionary where the keys are the type of metric and the values are lists of those metrics? We could also move the concern of instantiating the metrics outside of the registration methods.

That could then help us generalize some of the other code in this class while still ensuring we can do metric-type level manipulations in the future

```python3
def register_metric(
        metric: Metric
    ) -> None:
        self.metrics[type(metric)].append(metric)

def unregister_metric(
        metric: Metric
    ) -> None:
       self.metrics[type(metric)].remove(metric)

def get_overall_sample_size(self) -> float:
    metrics = reduce(operator.add, self.metrics.values())
    return max(map(self._get_single_sample_size, metrics))
```

Reply:
I wonder if Boolean metric require TTestIndPower when we support multiple metrics per calculator. We are going to modify this method anyway after we know the correction implementation we are going to use. I will keep this comment unresolved as a reference